### PR TITLE
Move downsample from tt_eager to ttnn

### PR DESCRIPTION
**Pull request by nilaykp**
### Ticket
[9760](https://github.com/tenstorrent/tt-metal/issues/9760)

### Problem description
Move downsample from tt_eager to ttnn

### What's changed
Move downsample cpp implementation ttnn, pybindings to ttnn and make respective documentation changes

### Checklist
- [x] Post commit CI passes [LINK](https://github.com/tenstorrent/tt-metal/actions/runs/9936579340)
- [x] [post-commit] models tests [LINK](https://github.com/tenstorrent/tt-metal/actions/runs/9936592205)
- [x] Device perf regressions and output report [LINK](https://github.com/tenstorrent/tt-metal/actions/runs/9911357460) 
- [x] Model perf regressions and output report [LINK](https://github.com/tenstorrent/tt-metal/actions/runs/9936662031)
- [ ] Nightly fast dispatch model regression CI [LINK](https://github.com/tenstorrent/tt-metal/actions/runs/9936671266)
- [x] T3000 frequent tests [LINK](https://github.com/tenstorrent/tt-metal/actions/runs/9936583829)
- [x] T3000 demo tests [LINK](https://github.com/tenstorrent/tt-metal/actions/runs/9936681186)

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/conv2d.cpp [hunk 1]
```diff
@@ -1,10 +1,10 @@
 // SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
-//
+//conv2d.cpp
```
Comment (TT-BrianLiu) [inline ttnn/cpp/ttnn/operations/conv2d.cpp:2]: Remove this?

**Review by TT-BrianLiu — APPROVED** [review]
Minor comment

File: ttnn/cpp/ttnn/operations/conv2d.cpp (hunk 1 above)
Comment (nkpatel-tt) [reply ttnn/cpp/ttnn/operations/conv2d.cpp]: Done

**Review by eyonland — APPROVED** [review]
Are you moving the tests on a separate PR?

**Comment by nkpatel-tt** [comment]
> Are you moving the tests on a separate PR?

The

> Are you moving the tests on a separate PR?

File: ttnn/cpp/pybind11/operations/__init__.hpp [hunk 2]
```diff
@@ -69,6 +70,9 @@ void py_module(py::module& module) {
     auto m_data_movement = module.def_submodule("data_movement", "data_movement operations");
     data_movement::py_module(m_data_movement);
 
+    auto m_downsample = module.def_submodule("downsample", "downsample operations");
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/pybind11/operations/__init__.hpp:73]: We don't need a dedicated submodule for downsample. Upsample and downsample ops should be in the same submodule

Comment (ayerofieiev-tt) [reply ttnn/cpp/pybind11/operations/__init__.hpp:73]: probably best belongs to conv

File: ttnn/cpp/ttnn/operations/conv2d.cpp [hunk 3]
```diff
@@ -668,7 +668,7 @@ std::tuple<ttnn::Tensor, uint32_t, uint32_t, ttnn::Tensor, std::optional<ttnn::T
         Tensor matmul_input = input_tensor_post_tm;
         if (stride[0] > 1) {
             // run downsample
-            matmul_input = tt::tt_metal::downsample(
+            matmul_input = ttnn::operations::data_movement::downsample(
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/conv2d.cpp:671]: It should not be in data_movement namespace. I think it belongs to convolutions, right?

File: ttnn/cpp/ttnn/operations/downsample/device/downsample_op.hpp [hunk 4]
```diff
@@ -25,6 +28,21 @@ struct Downsample {
     const auto attribute_values() const {
         return std::make_tuple(std::cref(this->downsample_params), std::cref(this->output_dtype));
     }
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/downsample/device/downsample_op.hpp:30]: attribute_names and attribute_values can be removed
See https://github.com/tenstorrent/tt-metal/commit/a0dbfe07f6726421ea90eeb02fe3548b9e5fee83

File: ttnn/cpp/ttnn/operations/downsample/device/downsample_op.hpp [hunk 5]
```diff
@@ -25,6 +28,21 @@ struct Downsample {
...
+    static ttnn::Tensor execute_on_worker_thread(
+        const Tensor& input_tensor_a, std::array<uint32_t, 5> downsample_params, std::optional<DataType> output_dtype) {
+        std::vector<Tensor> output_tensors = {Tensor(operation::get_workers_for_op_output({input_tensor_a}))};
+        operation::launch_op(
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/downsample/device/downsample_op.hpp:34]: This should not use `launch_op`.

Please switch to `operation::run`

File: ttnn/cpp/ttnn/operations/downsample/downsample_op_pybind.hpp [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+     )doc";
+
+    ttnn::bind_registered_operation(
+        module,
+        ttnn::downsample,
+        doc,
+        ttnn::pybind_arguments_t{
+            py::arg("input_tensor"),
+            py::arg("downsample_params"),
+            py::arg("output_dtype") = std::nullopt});
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/downsample/downsample_op_pybind.hpp:35]:
```suggestion
            py::arg("dtype") = std::nullopt});
```

File: ttnn/cpp/ttnn/operations/downsample/device/downsample_op.hpp [hunk 7]
```diff
@@ -36,6 +54,10 @@ Tensor downsample (const Tensor &a, std::array<uint32_t, 5> downsample_params, s
 // uint32_t get_num_cores(CoreCoord grid_size, uint32_t nblocks);
 // }
 
-}  // namespace tt_metal
+} // namespace data_movement
+
+}  // namespace operations
+
+constexpr auto downsample = ttnn::register_operation<ttnn::operations::data_movement::Downsample>("ttnn::downsample");
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/downsample/device/downsample_op.hpp:61]: You are mixing <Operation> and Execute<Operation> structures in one. Please follow the high level structure.
struct Downsample should contain all these 
```
    std::array<uint32_t, 5> downsample_params;
    DataType output_dtype;
    void validate(const std::vector<Tensor> &input_tensors) const;
    std::vector<tt::tt_metal::Shape> compute_output_shapes(const std::vector<Tensor> &input_tensors) const;
    std::vector<Tensor> create_output_tensors(const std::vector<Tensor> &input_tensors) const;
    operation::ProgramWithCallbacks create_program(const std::vector<Tensor>& input_tensors, std::vector<Tensor> &output_tensors) const;
```

but not execute_on_worker/main_thread


Please make yourself familiar with 
https://tenstorrent.github.io/tt-metal/latest/ttnn/ttnn/adding_new_ttnn_operation.html

CC @arakhmati

File: ttnn/cpp/ttnn/operations/downsample/downsample_op_pybind.hpp (hunk 6 above)
Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/downsample/downsample_op_pybind.hpp:35]: This op does not support optional memory_config and output_tensor?

**Review by tt-rkim — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/conv2d.cpp (hunk 3 above)
Comment (nkpatel-tt) [reply ttnn/cpp/ttnn/operations/conv2d.cpp:671]: @ayerofieiev-tt  upsample is part of data_movement. so shall I keep it in data_movement or you want me to move it to conv only??

Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/conv2d.cpp:671]: Does upsample really semantically belong to data_movement or its just that someone put it into that namespace? :)

File: ttnn/cpp/ttnn/operations/downsample/downsample_op_pybind.hpp (hunk 6 above)
Comment (nkpatel-tt) [reply ttnn/cpp/ttnn/operations/downsample/downsample_op_pybind.hpp:35]: I do not see them used explicitly. I think required memory config is calculated from input tensor only.

File: tests/tt_eager/python_api_testing/unit_testing/misc/test_downsample.py [hunk 8]
```diff
@@ -145,7 +146,7 @@ def test_run_downsample(
         ttl.tensor.TensorMemoryLayout.HEIGHT_SHARDED, ttl.tensor.BufferType.L1
     )
     # Run downsample op
-    A_downampled_sharded = ttl.tensor.downsample(A_sharded, downsample_params, output_dtype=dtype)
+    A_downampled_sharded = ttnn.downsample(A_sharded, downsample_params, output_dtype=dtype)
```
[the remark this answers is not in the document]
Comment (nkpatel-tt) [reply tests/tt_eager/python_api_testing/unit_testing/misc/test_downsample.py]: Done.

File: tt_eager/tt_lib/fused_ops/conv.py [hunk 9]
```diff
@@ -449,7 +450,7 @@ def resnet50_1x1_conv_s2_as_downsample_and_matmul(
 
     def conv_(activation):
         # downsample op
-        output = tensor.downsample(activation, downsample_params, output_dtype=output_dtype)
+        output = ttnn.downsample(activation, downsample_params, output_dtype=output_dtype)
```
[the remark this answers is not in the document]
Comment (nkpatel-tt) [reply tt_eager/tt_lib/fused_ops/conv.py]: Done.

File: ttnn/cpp/pybind11/operations/__init__.hpp (hunk 2 above)
Comment (nkpatel-tt) [reply ttnn/cpp/pybind11/operations/__init__.hpp]: kept in data_movement as per discussion.

Comment (nkpatel-tt) [reply ttnn/cpp/pybind11/operations/__init__.hpp]: Done.

File: ttnn/cpp/ttnn/operations/conv2d.cpp (hunk 3 above)
Comment (nkpatel-tt) [reply ttnn/cpp/ttnn/operations/conv2d.cpp:698]: kept in data_movement as of now.

File: ttnn/cpp/ttnn/operations/downsample/device/downsample_op.hpp (hunk 4 above)
Comment (nkpatel-tt) [reply ttnn/cpp/ttnn/operations/downsample/device/downsample_op.hpp]: Done.

File: ttnn/cpp/ttnn/operations/downsample/device/downsample_op.hpp (hunk 5 above)
Comment (nkpatel-tt) [reply ttnn/cpp/ttnn/operations/downsample/device/downsample_op.hpp]: Done.

File: ttnn/cpp/ttnn/operations/downsample/device/downsample_op.hpp (hunk 7 above)
Comment (nkpatel-tt) [reply ttnn/cpp/ttnn/operations/downsample/device/downsample_op.hpp]: Done.

File: ttnn/cpp/ttnn/operations/data_movement/downsample/downsample_op_pybind.hpp [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+namespace ttnn::operations::data_movement::detail {
+
+void bind_downsample(py::module& module) {
+    const auto doc = R"doc(
+ Downsamples a given multi-channel 2D (spatial) data.
+ The input data is assumed to be of the form [N, H, W, C].
+
+ Args:
+     * :attr:`input_tensor`: the input tensor
+     * :attr:`downsample_params`: Params list: batch size, conv input H, conv input W, conv stride H, conv stride W
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/downsample/downsample_op_pybind.hpp:27]: what Type is this Params?

Comment (nkpatel-tt) [reply ttnn/cpp/ttnn/operations/data_movement/downsample/downsample_op_pybind.hpp:27]: It is list in python.

**Pull request merged**